### PR TITLE
Fast Metal Tensor Loading

### PR DESCRIPTION
**Pull request by SkullMag**
I used the code below to evaluate the speed:
```python
import time
from tinygrad.tensor import Tensor
from tinygrad.helpers import getenv
import numpy as np

N = getenv("N", 10000)

timings = []
tensors = []
# Create random tensors
for i in range(10):
  data = np.random.default_rng().standard_normal(size=(N,N), dtype=np.float32) 
  tensors.append(Tensor(data, device=f"DISK:test{i}.data").realize())

# Move them to METAL
for t in tensors:
  s = time.monotonic()
  t.to("METAL").realize()
  timings.append(time.monotonic() - s)

tm = min(timings)   
print("Min timing:", tm)
```

Without `loadBuffer`
```sh
DEBUG=2 METAL=1 N=10000 LOAD_BUFFER=0 python test.py
***    1 copy DISK:test0.data <- CPU           arg   2 sz 100000000          dev DISK:test0.data OPs      0M/   0.00G  mem  0.00 GB tm 803258.42us/   803.26ms (    0.00 GFLOPS,    0.50 GB/s)
***    2 copy DISK:test1.data <- CPU           arg   2 sz 100000000          dev DISK:test1.data OPs      0M/   0.00G  mem  0.00 GB tm 689929.83us/  1493.19ms (    0.00 GFLOPS,    0.58 GB/s)
***    3 copy DISK:test2.data <- CPU           arg   2 sz 100000000          dev DISK:test2.data OPs      0M/   0.00G  mem  0.00 GB tm 442978.92us/  1936.17ms (    0.00 GFLOPS,    0.90 GB/s)
***    4 copy DISK:test3.data <- CPU           arg   2 sz 100000000          dev DISK:test3.data OPs      0M/   0.00G  mem  0.00 GB tm 131667.38us/  2067.83ms (    0.00 GFLOPS,    3.04 GB/s)
***    5 copy DISK:test4.data <- CPU           arg   2 sz 100000000          dev DISK:test4.data OPs      0M/   0.00G  mem  0.00 GB tm 139279.38us/  2207.11ms (    0.00 GFLOPS,    2.87 GB/s)
***    6 copy DISK:test5.data <- CPU           arg   2 sz 100000000          dev DISK:test5.data OPs      0M/   0.00G  mem  0.00 GB tm 334202.08us/  2541.32ms (    0.00 GFLOPS,    1.20 GB/s)
***    7 copy DISK:test6.data <- CPU           arg   2 sz 100000000          dev DISK:test6.data OPs      0M/   0.00G  mem  0.00 GB tm  34456.13us/  2575.77ms (    0.00 GFLOPS,   11.61 GB/s)
***    8 copy DISK:test7.data <- CPU           arg   2 sz 100000000          dev DISK:test7.data OPs      0M/   0.00G  mem  0.00 GB tm  52588.25us/  2628.36ms (    0.00 GFLOPS,    7.61 GB/s)
***    9 copy DISK:test8.data <- CPU           arg   2 sz 100000000          dev DISK:test8.data OPs      0M/   0.00G  mem  0.00 GB tm  47230.33us/  2675.59ms (    0.00 GFLOPS,    8.47 GB/s)
***   10 copy DISK:test9.data <- CPU           arg   2 sz 100000000          dev DISK:test9.data OPs      0M/   0.00G  mem  0.00 GB tm 144544.88us/  2820.14ms (    0.00 GFLOPS,    2.77 GB/s)
***   11 copy METAL   <- DISK:test0.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm  50538.00us/  2870.67ms (    0.00 GFLOPS,    7.91 GB/s)
***   12 copy METAL   <- DISK:test1.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm  38148.67us/  2908.82ms (    0.00 GFLOPS,   10.49 GB/s)
***   13 copy METAL   <- DISK:test2.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm 239841.08us/  3148.66ms (    0.00 GFLOPS,    1.67 GB/s)
***   14 copy METAL   <- DISK:test3.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm 102301.29us/  3250.96ms (    0.00 GFLOPS,    3.91 GB/s)
***   15 copy METAL   <- DISK:test4.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm  86443.08us/  3337.41ms (    0.00 GFLOPS,    4.63 GB/s)
***   16 copy METAL   <- DISK:test5.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm  36758.13us/  3374.17ms (    0.00 GFLOPS,   10.88 GB/s)
***   17 copy METAL   <- DISK:test6.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm  39298.17us/  3413.46ms (    0.00 GFLOPS,   10.18 GB/s)
***   18 copy METAL   <- DISK:test7.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm  37119.75us/  3450.58ms (    0.00 GFLOPS,   10.78 GB/s)
***   19 copy METAL   <- DISK:test8.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm 155120.17us/  3605.70ms (    0.00 GFLOPS,    2.58 GB/s)
***   20 copy METAL   <- DISK:test9.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm  36711.37us/  3642.42ms (    0.00 GFLOPS,   10.90 GB/s)
Min timing: 0.036959540999999874
avg:     0.00 GFLOPS     2.20 GB/s           total:    20 kernels     0.00 GOPS     8.00 GB  3642.42 ms
```

With `loadBuffer`

```sh
DEBUG=2 METAL=1 N=10000 LOAD_BUFFER=1 python test.py
***    1 copy DISK:test0.data <- CPU           arg   2 sz 100000000          dev DISK:test0.data OPs      0M/   0.00G  mem  0.00 GB tm 557979.12us/   557.98ms (    0.00 GFLOPS,    0.72 GB/s)
***    2 copy DISK:test1.data <- CPU           arg   2 sz 100000000          dev DISK:test1.data OPs      0M/   0.00G  mem  0.00 GB tm 828466.87us/  1386.45ms (    0.00 GFLOPS,    0.48 GB/s)
***    3 copy DISK:test2.data <- CPU           arg   2 sz 100000000          dev DISK:test2.data OPs      0M/   0.00G  mem  0.00 GB tm 186380.29us/  1572.83ms (    0.00 GFLOPS,    2.15 GB/s)
***    4 copy DISK:test3.data <- CPU           arg   2 sz 100000000          dev DISK:test3.data OPs      0M/   0.00G  mem  0.00 GB tm 120124.62us/  1692.95ms (    0.00 GFLOPS,    3.33 GB/s)
***    5 copy DISK:test4.data <- CPU           arg   2 sz 100000000          dev DISK:test4.data OPs      0M/   0.00G  mem  0.00 GB tm  50489.08us/  1743.44ms (    0.00 GFLOPS,    7.92 GB/s)
***    6 copy DISK:test5.data <- CPU           arg   2 sz 100000000          dev DISK:test5.data OPs      0M/   0.00G  mem  0.00 GB tm  34515.50us/  1777.96ms (    0.00 GFLOPS,   11.59 GB/s)
***    7 copy DISK:test6.data <- CPU           arg   2 sz 100000000          dev DISK:test6.data OPs      0M/   0.00G  mem  0.00 GB tm 258582.92us/  2036.54ms (    0.00 GFLOPS,    1.55 GB/s)
***    8 copy DISK:test7.data <- CPU           arg   2 sz 100000000          dev DISK:test7.data OPs      0M/   0.00G  mem  0.00 GB tm  33990.33us/  2070.53ms (    0.00 GFLOPS,   11.77 GB/s)
***    9 copy DISK:test8.data <- CPU           arg   2 sz 100000000          dev DISK:test8.data OPs      0M/   0.00G  mem  0.00 GB tm  33831.33us/  2104.36ms (    0.00 GFLOPS,   11.82 GB/s)
***   10 copy DISK:test9.data <- CPU           arg   2 sz 100000000          dev DISK:test9.data OPs      0M/   0.00G  mem  0.00 GB tm  86426.25us/  2190.79ms (    0.00 GFLOPS,    4.63 GB/s)
***   11 copy METAL   <- DISK:test0.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm    350.04us/  2191.14ms (    0.00 GFLOPS, 1142.72 GB/s)
***   12 copy METAL   <- DISK:test1.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm    112.96us/  2191.25ms (    0.00 GFLOPS, 3541.14 GB/s)
***   13 copy METAL   <- DISK:test2.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm    155.75us/  2191.41ms (    0.00 GFLOPS, 2568.22 GB/s)
***   14 copy METAL   <- DISK:test3.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm    226.04us/  2191.63ms (    0.00 GFLOPS, 1769.59 GB/s)
***   15 copy METAL   <- DISK:test4.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm    351.04us/  2191.98ms (    0.00 GFLOPS, 1139.46 GB/s)
***   16 copy METAL   <- DISK:test5.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm    521.50us/  2192.50ms (    0.00 GFLOPS,  767.02 GB/s)
***   17 copy METAL   <- DISK:test6.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm    254.79us/  2192.76ms (    0.00 GFLOPS, 1569.91 GB/s)
***   18 copy METAL   <- DISK:test7.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm    182.00us/  2192.94ms (    0.00 GFLOPS, 2197.80 GB/s)
***   19 copy METAL   <- DISK:test8.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm    154.04us/  2193.09ms (    0.00 GFLOPS, 2596.69 GB/s)
***   20 copy METAL   <- DISK:test9.data       arg   2 sz 100000000          dev METAL   OPs      0M/   0.00G  mem  0.40 GB tm    116.96us/  2193.21ms (    0.00 GFLOPS, 3420.03 GB/s)
Min timing: 0.0003297079999997621
avg:     0.00 GFLOPS     3.65 GB/s           total:    20 kernels     0.00 GOPS     8.00 GB  2193.21 ms
```

### CODE REVIEW

**Comment by imaolo** [comment]
[external](https://github.com/tinygrad/tinygrad/blob/master/test/external/external_benchmark_load_stable_diffusion.py)

File: tinygrad/runtime/ops_metal.py [hunk 1]
```diff
@@ -71,13 +72,23 @@ def as_buffer(self, src:Any) -> memoryview:
...
+    cbuf.commit()
 
 class MetalDevice(Compiled):
   compiler_device = None
   def __init__(self, device:str):
     self.device = Metal.MTLCreateSystemDefaultDevice()
     if MetalDevice.compiler_device is None: MetalDevice.compiler_device = self.device
     self.mtl_queue = self.device.newCommandQueueWithMaxCommandBufferCount_(1024)
+    self.mtl_io_queue, e = self.device.newIOCommandQueueWithDescriptor_error_(Metal.MTLIOCommandQueueDescriptor.new(), None)
+    self.mtl_io_command_buffer = self.mtl_io_queue.commandBuffer()
```
Comment (geohot) [inline tinygrad/runtime/ops_metal.py:90]: Why is this here?

File: tinygrad/runtime/ops_metal.py [hunk 2]
```diff
@@ -71,13 +72,23 @@ def as_buffer(self, src:Any) -> memoryview:
...
+    cbuf.loadBuffer_offset_size_sourceHandle_sourceHandleOffset_(dest._buf, 0, dest.size*dest.dtype.itemsize, handles[path], src._buf.offset)
+    cbuf.commit()
 
 class MetalDevice(Compiled):
   compiler_device = None
   def __init__(self, device:str):
     self.device = Metal.MTLCreateSystemDefaultDevice()
     if MetalDevice.compiler_device is None: MetalDevice.compiler_device = self.device
     self.mtl_queue = self.device.newCommandQueueWithMaxCommandBufferCount_(1024)
+    self.mtl_io_queue, e = self.device.newIOCommandQueueWithDescriptor_error_(Metal.MTLIOCommandQueueDescriptor.new(), None)
```
Comment (geohot) [inline tinygrad/runtime/ops_metal.py:89]: Use unwrap2

File: tinygrad/device.py [hunk 3]
```diff
@@ -103,6 +103,8 @@ def _internal_buffer_copy(dest, src):
     if fb:
       dest.allocator.transfer(dest._buf, fb, dest.size*dest.dtype.itemsize)
       return
+  if getenv("LOAD_BUFFER") and hasattr(dest.allocator, 'load_buffer') and src.device.startswith("DISK"):
```
Comment (geohot) [inline tinygrad/device.py:106]: Does this have to be gated with an env var? Is there ever a downside?

**Comment by geohot** [comment]
Screenshot 2023-12-10 at 5 41 11 PM

broken

File: tinygrad/runtime/ops_metal.py [hunk 4]
```diff
@@ -71,13 +72,23 @@ def as_buffer(self, src:Any) -> memoryview:
     return src.contents().as_buffer(src.length())
   def copyin(self, dest:Any, src:memoryview): self.as_buffer(dest)[:] = src
   def copyout(self, dest:memoryview, src:Any): dest[:] = self.as_buffer(src)
+  def load_buffer(self, dest:Any, src:Any, handles={}):
+    if (path := src.device.split(":")[1]) not in handles:
+      handles[path], err = self.device.device.newIOHandleWithURL_error_(NSURL.fileURLWithPath_(path), None)
+      assert err is None, "failed to load from disk"
+    cbuf = self.device.mtl_io_queue.commandBuffer()
+    cbuf.loadBuffer_offset_size_sourceHandle_sourceHandleOffset_(dest._buf, 0, dest.size*dest.dtype.itemsize, handles[path], src._buf.offset)
```
Comment (geohot) [inline tinygrad/runtime/ops_metal.py:80]: The wrong type is being passed in here

**Comment by geohot** [comment]
I locked the bounty for you. Can you test examples/coder.py?

File: tinygrad/device.py (hunk 3 above)
Comment (SkullMag) [reply tinygrad/device.py:109]: it doesn't load tensor synchronously so you have to be aware of syncing device before usage

Comment (nullhook) [reply tinygrad/device.py:109]: according to Metal's feature set table, MTLIO is supported everywhere. so i'm not sure if a gate is required here.

image

Comment (SkullMag) [reply tinygrad/device.py]: removed that. it was for testing purposes

**Comment by SkullMag** [comment]
> Screenshot 2023-12-10 at 5 41 11 PM
> 
> broken, how is this being synced?

It's synced rn. Forgot to add those buffers to the list of in-flight buffers

File: tinygrad/runtime/ops_metal.py (hunk 4 above)
Comment (SkullMag) [reply tinygrad/runtime/ops_metal.py:78]: dest._buf - metal buffer
0 - int
dest.size*dest.dtype.itemsize - int
handles[path] - mtliofilehandle
src._buf.offset - int

I guess everything is right

**Comment by SkullMag** [comment]
So, after making `DiskAllocator` lazy (open file only when `UnderlyingDiskBuffer` is first accessed), I was able to get those results:

```
sudo purge # This command was run before each benchmark
```
current master:
```
ram used:  4.26 GB, cond_stage_model.transformer.text_model.final_layer_norm.bias: 100%|█████████████████████████████████| 1131/1131 [00:04<00:00, 251.63it/s]
loaded weights in 4506.39 ms, 4.26 GB loaded at 0.95 GB/s
4669.73 ms
```

this branch:
```
ram used:  4.26 GB, cond_stage_model.transformer.text_model.final_layer_norm.bias: 100%|█████████████████████████████████| 1131/1131 [00:02<00:00, 466.11it/s]
loaded weights in 2438.77 ms, 4.26 GB loaded at 1.75 GB/s
2596.40 ms
```

**Comment by geohot** [comment]
What computer is that benchmark on? Are you purging before? How does this impact performance if you don't purge?

**Comment by geohot** [comment]
Tried this today:
```
diane@blak:~/tinygrad$ python3 examples/llama.py --prompt "Hello." --count 10 --temperature 0 --timing --gen 2
using METAL backend
using LLaMA-2-7B model
ram used: 13.48 GB, freqs_cis                                         : 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 292/292 [00:01<00:00, 214.47it/s]
loaded weights in 1368.95 ms, 13.48 GB loaded at 9.84 GB/s
Hello.
ran model in 405.80 ms
total 470.41 ms, 2.13 tok/sec
ilde
ran model in 311.57 ms
total 313.63 ms, 3.19 tok/sec
ilde
ran model in 243.99 ms
total 246.45 ms, 4.06 tok/sec
ilde
ran model in 2.09 ms
total 53.36 ms, 18.74 tok/sec
ilde
ran model in 2.08 ms
total 48.34 ms, 20.69 tok/sec
ilde
ran model in 1.74 ms
total 47.99 ms, 20.84 tok/sec
ilde
ran model in 2.39 ms
total 47.97 ms, 20.85 tok/sec
ilde
ran model in 2.86 ms
total 48.05 ms, 20.81 tok/sec
ilde
ran model in 2.71 ms
total 48.25 ms, 20.72 tok/sec
ilde
ran model in 3.73 ms
total 48.58 ms, 20.59 tok/sec
ildeTraceback (most recent call last):
  File "/Users/diane/tinygrad/examples/llama.py", line 445, in <module>
    assert text == expected[key], "invalid output: " + colored(text, "red")
           ^^^^^^^^^^^^^^^^^^^^^
AssertionError: invalid output: Hello.ildeildeildeildeildeildeildeildeildeilde
```

File: tinygrad/runtime/ops_metal.py [hunk 5]
```diff
@@ -60,23 +61,26 @@ def transfer(self, dest:Any, src:Any, sz:int):
...
-    ret = self.device.device.newBufferWithBytesNoCopy_length_options_deallocator_(src, len(src), Metal.MTLResourceStorageModeShared, None)
-    if ret: self.device.mv_in_metal.append(src)
-    return ret
   def _free(self, opaque:Any): opaque.release()
   def as_buffer(self, src:Any) -> memoryview:
     self.device.synchronize()
     return src.contents().as_buffer(src.length())
   def copyin(self, dest:Any, src:memoryview): self.as_buffer(dest)[:] = src
   def copyout(self, dest:memoryview, src:Any): dest[:] = self.as_buffer(src)
+  def load_buffer(self, dest:Any, src:Any, handles:Dict={}):
```
Comment (geohot) [inline tinygrad/runtime/ops_metal.py:70]: While this is clever, this should be a property added to the ud buffer.

**Comment by geohot** [comment]
"You can synchronize these queues with a Metal shared event. Metal hared events let you synchronize the command buffers from your IO queue with the command buffers from your rendering queue. You can tell a command buffer to wait for a shared event by encoding a waitEvent command."

**Comment by SkullMag** [comment]
I benchmarked it on Macbook air m1, 16 GB RAM, 1 TB SSD. Without purging speed is virtually the same. Disk read speed is reported to be ~3000 MB/s

**Comment by SkullMag** [comment]
I don't see a speed improvement here

File: tinygrad/runtime/ops_metal.py (hunk 5 above)
Comment (SkullMag) [reply tinygrad/runtime/ops_metal.py:70]: no, they are not closed rn. i'll fix that if I see a speed increase of load_buffer
